xprof/btx_timeline.cpp

```diff
@@ -72,78 +78,122 @@ static perfetto_uuid_t get_parent_counter_track_uuid(timeline_dispatch_t *dispat
+  oss << "Hostname " << hostname << " | Device " << deviceIdx;
   // oss << " | " << track_name << " | uuid ";
   process->set_process_name(oss.str());
   return hp_uuid;
 }
 
-static perfetto_uuid_t
-get_counter_track_uuuid(timeline_dispatch_t *dispatch,
-                        std::unordered_map<hp_ddomain_t, perfetto_uuid_t> &counter_tracks,
-                        const std::string track_name, std::string hostname, uint64_t process_id,
-                        thapi_device_id did, thapi_domain_idx domain, int64_t unit_multiplier = 1) {
+static perfetto_uuid_t get_counter_track_uuuid(timeline_dispatch_t *dispatch,
+                                               std::unordered_map<hp_ddomain_t, perfetto_uuid_t> &counter_tracks,
+                                               const std::string &track_name, const std::string &hostname, uint64_t process_id,
+                                               thapi_device_id did, uint32_t deviceIdx, uint64_t tHandle, thapi_domain_idx domain,
+                                               std::optional<Details> details = std::nullopt,
+                                               std::unordered_map<hp_dfsdev_t, perfetto_uuid_t> *counter_tracks_fp = nullptr) {
   perfetto_uuid_t hp_dev_uuid = 0;
-  auto [it, inserted] = counter_tracks.insert({{hostname, process_id, did, domain}, hp_dev_uuid});
-  auto &potential_uuid = it->second;
-  // Exists
-  if (!inserted)
-    return potential_uuid;
-
-  perfetto_uuid_t hp_uuid = get_parent_counter_track_uuid(dispatch, hostname, process_id, did);
-  hp_dev_uuid = gen_perfetto_uuid();
-  potential_uuid = hp_dev_uuid;
+  perfetto_uuid_t hp_uuid;
+
+    if (details && counter_tracks_fp != nullptr) {
+    auto [it, inserted] = counter_tracks_fp->insert({{hostname, process_id, did, tHandle, domain,  details->RxTx},
+                                                    hp_dev_uuid});
+    auto &potential_uuid = it->second;
+    if (!inserted)
+      return potential_uuid;
+
+    hp_uuid = get_parent_counter_track_uuid(dispatch, hostname, process_id, did, deviceIdx);
+    hp_dev_uuid = gen_perfetto_uuid();
+    potential_uuid = hp_dev_uuid;
+  } else {
+    auto [it, inserted] = counter_tracks.insert({{hostname, process_id, did, domain, tHandle}, hp_dev_uuid});
+    auto &potential_uuid = it->second;
+    if (!inserted)
+      return potential_uuid;
+
+    hp_uuid = get_parent_counter_track_uuid(dispatch, hostname, process_id, did, deviceIdx);
+    hp_dev_uuid = gen_perfetto_uuid();
+    potential_uuid = hp_dev_uuid;
+  }
 
-  // Create new track
   auto *packet = dispatch->trace.add_packet();
   packet->set_timestamp(0);
   packet->set_trusted_packet_sequence_id(TRUSTED_PACKED_SEQUENCE_ID);
   auto *track_descriptor = packet->mutable_track_descriptor();
   track_descriptor->set_uuid(hp_dev_uuid);
   track_descriptor->set_parent_uuid(hp_uuid);
   std::ostringstream oss;
-  oss << track_name << " | Domain " << domain;
+  if (track_name=="FabricT" && details) {
```

Comment:
So we cannot ever use `FabricT` without `details`?

Reply:
Yes, in our context we are using details for Fabric and Memory stats.  We wanted not to lose the remote and the fabric Ids on the timeline. In addition, the direction of the data movement. In the case of memory stats we need the direction of the movement. These are not available on the other metrics so, to avoid  redundancy, I used details as optional for the two.